core/src/main/java/org/jdbi/v3/core/inference/JdbiInterceptorChain.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.inference;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * An interception chain to process transformation operations.
+ *
+ * @param <S> The source type.
+ * @param <T> The target type.
+ */
+public final class JdbiInterceptorChain<S, T> {
+
+    private final List<JdbiInterceptor<S, T>> interceptors = new CopyOnWriteArrayList<>();
+    private JdbiInterceptor<S, T> defaultInterceptor;
```

Comment:
I wanted to make sure that the default interceptor is always run last. With the "addFirst/addLast", one could otherwise not add an interceptor "to the end, right before the last one" unless we want to do length()-1 arithmetics when adding an interceptor. I am open to this; it would just be an implementation detail.